flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -158,7 +159,8 @@ public MemorySegment requestMemorySegment() {
 
     public List<MemorySegment> requestMemorySegmentsBlocking(int numberOfSegmentsToRequest)
             throws IOException {
-        return internalRequestMemorySegments(numberOfSegmentsToRequest);
+        return internalRequestMemorySegments(
+                numberOfSegmentsToRequest, this::internalRecycleMemorySegments);
```

Comment:
Why is it safe to acquire the lock in `requestMemorySegmentsBlocking` but no in `requestMemorySegments`?

As a side note, the names of those methods seems to be misleading, as both of them are blocking, and they only seem to differ with the "blocking" version not invoking `tryRedistributeBuffers()`?

Regardless of that, it seems to me like a fragile contract if two almost identical methods would be behaving in a quite different way. One would be acquiring a lock and potentially deadlocking, the other not.

Reply:
I agree that the method name is misleading. Currently, the ```requestMemorySegments``` method is only used by InputChannel for exclusive buffer requesting and it will add the number of required buffers to the total required buffers. If the total required buffers is larger than the total number of buffers, the "Insufficient number of network buffers" exception will be thrown. Because the logic number of available buffers (totalNumberOfMemorySegments - numTotalRequiredBuffers) is changed, buffer redistribution is needed. Different from the ```requestMemorySegments``` method, the ```requestMemorySegment``` and ```requestMemorySegmentsBlocking``` methods are only used by ```LocalBufferPool```, because ```LocalBufferPool``` has already reserved required buffers logically when it was created, so when requesting buffers, the logic number of available buffers is not changed, as a result, buffer redistribution is not needed. What do you think is I change the method name of ```requestMemorySegments``` to ```requestExclusiveSegments```? Any suggestions?

Reply:
Change the name of ```requestMemorySegments``` need to change the interface name of MemorySegmentProvider. Another choice maybe change the method name used by ```LocalBufferPool``` to something like ```pollSegment``` or ```pollSegmentsBlocking```.

Reply:
So the most important difference between `requestMemorySegments` (I) vs `requestMemorySegment` (II) and `requestMemorySegmentsBlocking` (III), is that buffers request by (II) or (III) are already accounted as "required" during the creation of the `LocalBufferPool`? While (I) are not previously accounted anywhere? Why is it that way? Why can not we always calculate the required buffers upon `LocalBufferPool` creation?

Reply:
Because the exclusive buffers used by remote channels are not allocated from LocalBufferPool. They are allocated directly from NetworkBufferPool using the requestMemorySegments (I) method. The LocalBufferPool allocates buffers lazily, so it accounts those buffers as "required" when initializing and then allocates on-demand. Instead, the exclusive buffers used by remote channels are allocated and accounted as "required" at the same time on remote channel setup.

Reply:
But why is done like so? Why can not we request exclusive buffers via `LocalBufferPool`? I see that it has been done like that since 2017, when credit based flow control was introduced.

I mean this is an independent issue from FLINK-25407, but as I see it we could:

1. either somehow rename those methods. But I have a feeling that every potential name will be confusing, until we find an answer to the above question, and embed this answer into the code (method name or a comment).
2. or we could maybe simplify contract of those methods/classes, and unify the way how exclusive buffers are allocated in the first place?

edit: 
> But why is done like so? Why can not we request exclusive buffers via `LocalBufferPool`?

Is it because all of the buffers in the `LocalBufferPool` are floating buffers? And that's the basis of the implementation of the whole exclusive/floating buffer concept?

Reply:
> Is it because all of the buffers in the LocalBufferPool are floating buffers? And that's the basis of the implementation of the whole exclusive/floating buffer concept?

I guess you are right, if we use only one LocalBufferPool, the buffers will be shared by all remote channels, but the credit based flow control  needs exclusive buffers (if we set the exclusive buffer to 0 and makes all buffers floating, some network micro benchmarks incurs at most about 20% regression). Another possible way is to create a local buffer pool per remote channel?

In a word, we can rename the method name, add some comments or try to unify the code to always use local buffer pool (need run micro benchmark to ensure that there is no regression). All of these are acceptable to me. WDYT, any suggestion?

Reply:
Maybe lets rename `requestMemorySegments` to something like `requestUnpooledMemorySegments()` (but please do so as a separate commit - can be in this PR), with the following java doc:
```
Unpooled memory segments are requested directly from {@link NetworkBufferPool}, as opposed to pooled segments, that are requested through {@link BufferPool} that was created from this {@link NetworkBufferPool} (see {@link NetworkBufferPool#createBufferPool(...)}). They are used for example for exclusive {@link RemoteInputChannel} credits, that are permanently assigned to that channel, and never returned to any {@link BufferPool}.
```

I don't like that "unpooled" is ambiguous, because we have two levels of pooling, but I don't know how to express it differently.

Reply:
I think your proposal is good enough. I will update the PR soon.